### PR TITLE
Integrate Cutlass Fused Multihead Attention in PHI

### PR DESCRIPTION
**Pull request by MARD1NO**
### PR types
New features

### PR changes
OPs

### Describe
Integrate Cutlass fused multihead attention

You can Add custom attention_mask

cutlass2.11.0兼容问题，参考 https://github.com/PaddlePaddle/Paddle/pull/50073#discussion_r1090194218 PR修改即可

文档：
image

### Benchmark
dev: cuda11.6 A100 40G

The case is borrowed from [xformers](https://github.com/facebookresearch/xformers/pull/587#issuecomment-1353245893) 

#### FP16: 
Without mask:
| |Cutlass Time(ms)|Naive Time(ms)|
|---|----|----|
| 32, 128, 16, 64 | 0.05 | 0.25 |
| 64, 128, 16, 16 | 0.07 | 0.27 |
| 64, 128, 16, 32 | 0.08 | 0.32 |
| 64, 512, 16, 16 | 0.91 | 2.75 |
| 64, 512, 16, 32 | 0.92 | 2.97 |
| 64, 512, 16, 64 | 1.04 | 3.36 |
| 64, 1024, 16, 128 | 6.62 | 18.31 |
| 64, 1024, 16, 256 | 15.36 | 22.52|

With mask:
| |Cutlass Time(ms)|Naive Time(ms)|
|---|----|----|
| 32, 128, 16, 64 | 0.09 | 0.29 |
| 64, 128, 16, 16 | 0.12 | 0.35 |
| 64, 128, 16, 32 | 0.13 | 0.4 |
| 64, 512, 16, 16 | 1.58 | 3.91 |
| 64, 512, 16, 32 | 1.58 | 4.12 |
| 64, 512, 16, 64 | 1.68 | 4.53 |
| 64, 1024, 16, 128 | 9.44 | 22.92 |
| 64, 1024, 16, 256 | 17.97 | 27.12 |

### InferCase FP16
| (b, q_seq, kv_seq, num_head, head_size) |Cutlass Time(us)|Naive Time(us)|
|---|----|----|
| 1, 900, 6000, 32 (PETR FMCA)| 376.51 | 625.1 |
| 1, 4096, 4096, 8, 32 (Diffusion FMHA)| 554.87 | 1494.8 |
| 1, 4096, 77, 8, 40 (Diffusion FMCA)| 34.81 | 119.14 |
| 1, 197, 197, 12, 64 (VIT FMHA)| 21.70 | 73.08 |

### Compare script
```python
def naive_attention_impl(query, key, value, mask, scale): 
    query = paddle.transpose(query, [0, 2, 1, 3])
    key = paddle.transpose(key, [0, 2, 1, 3])
    value = paddle.transpose(value, [0, 2, 1, 3])

    qk_res = paddle.matmul(query, key, transpose_y=True)
    attention = qk_res * scale
    attention = attention + mask 
    softmax_result = paddle.nn.functional.softmax(attention, -1)
    result = paddle.matmul(softmax_result, value)
    result = paddle.transpose(result, [0, 2, 1, 3])
    return result
```

### TODO List
generate.sh “借鉴”自xformers，通过shell脚本生成对应模板特化kernel，实现并行编译，加快编译速度

后续可以考虑采用python脚本来实现Kernel生成。

### CODE REVIEW

**Comment by mnicely** [comment]
@MARD1NO I wanted to make you aware of a change coming to our fMHA ([773](https://github.com/NVIDIA/cutlass/pull/773))

@hwu36 for vis

**Comment by MARD1NO** [comment]
> @MARD1NO I wanted to make you aware of a change coming to our fMHA ([773](https://github.com/NVIDIA/cutlass/pull/773))
> 
> @hwu36 for vis

Thanks, I will keep following this update

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -7,7 +7,8 @@ exclude: |
         python/paddle/utils/gast/.+|
         .+_pb2\.py|
         python/paddle/fluid/tests/unittests/npu/.+|
-        python/paddle/fluid/tests/unittests/mlu/.+
+        python/paddle/fluid/tests/unittests/mlu/.+|
+        paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/.+
```
Comment (MARD1NO) [inline .pre-commit-config.yaml:10]: 这里引入的是外部xformers代码，暂时不做format

File: python/paddle/fluid/tests/unittests/test_cutlass_fused_multihead_attention.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+import math
+import os
+import unittest
+
+import numpy as np
+from op_test import OpTest
+
+# Ensure we use float type to accumulate
+os.environ["FLAGS_gemm_use_half_precision_compute_type"] = "0"
```
Comment (MARD1NO) [inline python/paddle/fluid/tests/unittests/test_cutlass_fused_multihead_attention.py:23]: 保证对比的naive实现gemm使用float累加

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernels/generate_kernels.sh [hunk 3]
```diff
@@ -0,0 +1,38 @@
+# This generate kernel template script is referenced from 
+# https://github.com/facebookresearch/xformers/blob/main/xformers/csrc/attention/cuda/fmha/kernels/generate_kernels.sh
+
+#!/bin/bash
+set -ex
```
Comment (MARD1NO) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernels/generate_kernels.sh]: 这里参考使用xformers的算子模板生成脚本，以实现并行编译，加快速度

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernel_forward.h [hunk 4]
```diff
@@ -0,0 +1,1028 @@
...
+      }
+
+      typename MM0::Mma::Operator::IteratorC::TensorCoord
+          iteratorC_tile_offset = {
+              (tb_tile_offset.m() * MM0::Mma::WarpCount::kM) +
+                  (my_warp_id % MM0::Mma::WarpCount::kM),
+              (tb_tile_offset.n() * MM0::Mma::WarpCount::kN) +
+                  (my_warp_id / MM0::Mma::WarpCount::kM)};
+
+      if (kAddMask) {
```
Comment (MARD1NO) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernel_forward.h:622]: 如果要在QKmatmul后加mask，则需要将scale提前在寄存器计算好，而不是放到最后的tiledsoftmax里一起做

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernel_forward.h [hunk 5]
```diff
@@ -0,0 +1,1028 @@
...
+        accum =
+            cutlass::multiplies<typename MM0::Mma::FragmentC>()(p.scale, accum);
+      }
+
+      int32_t mask_iter_m = kMaskBroadcastRow ? 1 : problem_size_0_m;
```
Comment (MARD1NO) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernel_forward.h:627]: 这里对mask的行broadcast做了一个特化

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/gemm/custom_mma.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
+#pragma once
```
Comment (carryyu) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/gemm/custom_mma.h:1]: 确认下这个文件夹下的custom相关文件是否需要？

Comment (MARD1NO) [reply paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/gemm/custom_mma.h:1]: 现在forward这部分其实都放到了mma_from_smem.h了，但我看xformers仓库官方backward部分还是引用的

Comment (carryyu) [reply paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/gemm/custom_mma.h:1]: 从这个文件往下的四个custom文件应该是不需要的

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernel_forward.h [hunk 7]
```diff
@@ -0,0 +1,1028 @@
...
+      EpilogueOutputOp rescale(s_prime, m_prime);
+      Epilogue epilogue(
+          shared_storage.epilogue_shared_storage(),
+          thread_id(),
+          warp_id(),
+          lane_id());
+      epilogue(rescale, dest_iter, accum_o);
+    }
+
+    // 7. Calculate logsumexp
```
Comment (carryyu) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernel_forward.h:878]: 推理是不是不需要这一部分

Comment (MARD1NO) [reply paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernel_forward.h:879]: 是不需要，所以这里会判断logsumexp_ptr是否为nullptr，在Kernel的入口函数目前是直接设置为nullptr

Comment (MARD1NO) [reply paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernel_forward.h:878]: > 推理是不是不需要这一部分

是不需要，所以这里会判断logsumexp_ptr是否为nullptr，在Kernel的入口函数目前是直接设置为nullptr

**Comment by vivienfanghuagood** [comment]
LGTM

**Review by vivienfanghuagood — DISMISSED** [review]
LGTM

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/gemm/custom_mma_pipelined.h [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+    iterator_A.load(tb_frag_A);
+    iterator_B.load(tb_frag_B);
+
+    ++iterator_A;
+    ++iterator_B;
+
+    this->smem_iterator_A_.store(transform_A(tb_frag_A));
+    this->smem_iterator_B_.store(transform_B(tb_frag_B));
+
+    ++this->smem_iterator_A_;
```
Comment (zhoutianzi666) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/gemm/custom_mma_pipelined.h:285]: `iterator_B`是用来扫描global memory的，它应该不断++。
`smem_iterator_A_`我理解是用来写shared memory的，为啥需要++操作呢？

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+  Here M is: query_seq_len, K is: head_size, N is: key_value_seq_len.
+  The stride concept is equals to torch's, it means the offset to move to next
+  axis. For Q matrix(M, K), we need to move K(which equals to head_size) offset
+  to next row(in M axis).
+  */
+  int32_t query_strideH;
+  int32_t key_strideH;
+  int32_t value_strideH;
+
+  // Since mask can be broadcasted, we need to assign each stride
```
Comment (MARD1NO) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu:74]: 由于mask要支持各种broadacst case，所以各个维度的stride需要指定

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+
+  // Dimensions/strides
+  int32_t num_batches;
+  int32_t num_heads;
+  int32_t query_seq_len;
+  int32_t key_value_seq_len;
+  int32_t head_size;
+  int32_t value_head_size;
+  bool causal;
+  bool mask_broadcast_row;
```
Comment (MARD1NO) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu:60]: 是否在行方向上进行broadcast

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+    p.key_ptr = const_cast<T*>(reinterpret_cast<const T*>(params.key_ptr));
+    p.value_ptr = const_cast<T*>(reinterpret_cast<const T*>(params.value_ptr));
+    p.mask_ptr = const_cast<T*>(reinterpret_cast<const T*>(params.mask_ptr));
+
+    // TODO(zhengzekang): Currently we only support inference, so here we set
+    // `logsumexp_ptr` as nullptr, which is used for backward.
+    p.logsumexp_ptr = nullptr;
+
+    p.output_accum_ptr = nullptr;
+    if (Attention::kNeedsOutputAccumulatorBuffer) {
```
Comment (MARD1NO) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu:111]: 如果输出太大需要outputbuffer，需要临时申请一块

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/gemm/custom_mma_pipelined.h (hunk 8 above)
Comment (MARD1NO) [reply paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/gemm/custom_mma_pipelined.h:285]: （我的回答不一定准确）
这部分是mma_pipelined部分，是针对sm80以下架构显卡来做流水。sm80有一个async copy shared memory直接从global mem异步拷贝到smem，不需要额外寄存器作为中转，即可形成流水。而其他架构，只能开2倍的寄存器和smem来做流水（这也是为什么pipeline版本的stage是固定为2）。

这里++操作我觉得应该是移动到下一大块smem（因为你smem开了原来的两倍），到后面会有一个
```
this->smem_iterator_A_.add_tile_offset({0, -Base::kStages});
this->smem_iterator_B_.add_tile_offset({-Base::kStages, 0});
```
回退回去

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+  } else if (compute_capability == 75) {
+    DispatchFMHAIsAligned<T, cutlass::arch::Sm75>(params, ctx);
+  } else if (compute_capability == 70) {
+    DispatchFMHAIsAligned<T, cutlass::arch::Sm70>(params, ctx);
+  } else {
+    PADDLE_ENFORCE_EQ(true,
+                      false,
+                      phi::errors::Unimplemented(
+                          "Currently cutlass fused multihead attention kernel "
+                          "only support arch: SM80, SM75, SM70"));
```
Comment (wwbitejotunn) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu]: 这里是不是用PADDLE_THROW(platform::errors::Unimplemented());这种更清晰一些?

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/gemm/custom_mma_pipelined.h (hunk 8 above)
Comment (zhoutianzi666) [reply paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/gemm/custom_mma_pipelined.h:285]: > （我的回答不一定准确） 这部分是mma_pipelined部分，是针对sm80以下架构显卡来做流水。sm80有一个async copy shared memory直接从global mem异步拷贝到smem，不需要额外寄存器作为中转，即可形成流水。而其他架构，只能开2倍的寄存器和smem来做流水（这也是为什么pipeline版本的stage是固定为2）。
> 
> 这里++操作我觉得应该是移动到下一大块smem（因为你smem开了原来的两倍），到后面会有一个
> 
> ```
> this->smem_iterator_A_.add_tile_offset({0, -Base::kStages});
> this->smem_iterator_B_.add_tile_offset({-Base::kStages, 0});
> ```
> 
> 回退回去
有道理

**Review by carryyu — DISMISSED** [review]

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernels/forward.h [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+                                  MASK_BROADCAST_ROW>)      \
+  printf(                                                           \
+      "FATAL: this function is for sm%d, but was built for sm%d\n", \
+      int(ARCH),                                                    \
+      int(__CUDA_ARCH_OR_ZERO__));                                  \
+  _ATTENTION_KERNEL_FORWARD_END();
+
+// Enable the right one based on __CUDA_ARCH__
+#ifndef __CUDA_ARCH__
+if  __CUDA_ARCH__ && > 700__CUDA_ARCH__ < 750
```
Comment (carryyu) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernels/forward.h]: 写法需要修改，不然编译会报错

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu [hunk 14]
```diff
@@ -0,0 +1,393 @@
...
+
+template <typename T,
+          typename ArchTag,
+          bool IsAligned,
+          int QueriesPerBlock,
+          int KeysPerBlock>
+void DispatchFMHASingleValueIteration(LaunchParams params,
+                                      const phi::GPUContext& ctx) {
+  if (params.value_head_size <= KeysPerBlock ||
+      (KeysPerBlock == 64 && QueriesPerBlock == 64)) {
```
Comment (carryyu) [inline paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu]: 第二个判断条件是不是没有必要

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu (hunk 12 above)
Comment (MARD1NO) [reply paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu]: 已修改

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernels/forward.h (hunk 13 above)
Comment (MARD1NO) [reply paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention/kernels/forward.h]: 已修改

File: paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu (hunk 14 above)
Comment (MARD1NO) [reply paddle/phi/kernels/fusion/cutlass/fused_multi_head_attention.cu]: 已修改

**Comment by zhoutianzi666** [comment]
image

https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/incubate/nn/FusedMultiHeadAttention_cn.html#fusedmultiheadattention

- 这个API是不是应该去掉cutlass_fused_multi_head attention中的fused？

**Comment by MARD1NO** [comment]
> image
> 
> https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/incubate/nn/FusedMultiHeadAttention_cn.html#fusedmultiheadattention
> 
> * 这个API是不是应该去掉cutlass_fused_multi_head attention中的fused？

这里的fused是指 q matmul k, softmax, attention matmul v三个操作。原来的FusedMultiHeadAttention实质上也是拆成多个算子来完成操作，这个cutlass算子只启动了一个kernel完成

**Review by heavengate — DISMISSED** [review]

**Review by heavengate — APPROVED** [review]
